<a href="https://colab.research.google.com/github/Fowzia10/ETE456/blob/main/Copy_of_Task1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

CHITTAGONG UNIVERSITY OF ENGINEERING AND TECHNOLOGY

Name: Fowzia Hossain

ID :1608010

Course No.: ETE 456 

**Objective:**

1.Finding Charges of given dataset

*we use filterwarnings() function to supress warnings

In [ ]:
import warnings
warnings.filterwarnings("ignore")


*Imported numpy pandas and matlab library

In [ ]:
import numpy as np        
import pandas as pd     
import matplotlib.pyplot as plt       


Simple Linear Regression

In [ ]:
# Download the data
!wget -O insurance.csv https://www.dropbox.com/s/mwgqgjbmfw0xa5p/insurance.csv?dl=0


--2021-12-19 15:14:47--  https://www.dropbox.com/s/mwgqgjbmfw0xa5p/insurance.csv?dl=0
Resolving www.dropbox.com (www.dropbox.com)... 162.125.3.18, 2620:100:601b:18::a27d:812
Connecting to www.dropbox.com (www.dropbox.com)|162.125.3.18|:443... connected.
HTTP request sent, awaiting response... 301 Moved Permanently
Location: /s/raw/mwgqgjbmfw0xa5p/insurance.csv [following]
--2021-12-19 15:14:47--  https://www.dropbox.com/s/raw/mwgqgjbmfw0xa5p/insurance.csv
Reusing existing connection to www.dropbox.com:443.
HTTP request sent, awaiting response... 302 Found
Location: https://ucf596704303e38c668d7d577607.dl.dropboxusercontent.com/cd/0/inline/BcLmXOyF3SZo3qpWPgItTP0-PjNQExLnNK6pd8Jpd7PbcAMJjExAHMaU1jKnLPTMoCd-S4iRSWGAWYAbFAnHXQe6TLN5KorXSP1NwcD_VyQd0G4RXQz3OVeP3oeMdwaV7t-KrMnbUBjlhCx1DyqRt3Vm/file# [following]
--2021-12-19 15:14:48--  https://ucf596704303e38c668d7d577607.dl.dropboxusercontent.com/cd/0/inline/BcLmXOyF3SZo3qpWPgItTP0-PjNQExLnNK6pd8Jpd7PbcAMJjExAHMaU1jKnLPTMoCd-S4iRSWGAWYAbFA

In [ ]:

"""importing the dataset """
dataset = pd.read_csv('insurance.csv')
dataset


age     sex     bmi  children smoker     region      charges
0      19  female  27.900         0    yes  southwest  16884.92400
1      18    male  33.770         1     no  southeast   1725.55230
2      28    male  33.000         3     no  southeast   4449.46200
3      33    male  22.705         0     no  northwest  21984.47061
4      32    male  28.880         0     no  northwest   3866.85520
...   ...     ...     ...       ...    ...        ...          ...
1333   50    male  30.970         3     no  northwest  10600.54830
1334   18  female  31.920         0     no  northeast   2205.98080
1335   18  female  36.850         0     no  southeast   1629.83350
1336   21  female  25.800         0     no  southwest   2007.94500
1337   61  female  29.070         0    yes  northwest  29141.36030

[1338 rows x 7 columns]

In [ ]:
feature = dataset[['age','bmi','sex','children','smoker','region']]  # for independent
target = dataset[['charges']]   # for dependent


In [ ]:
feature

age     bmi     sex  children smoker     region
0      19  27.900  female         0    yes  southwest
1      18  33.770    male         1     no  southeast
2      28  33.000    male         3     no  southeast
3      33  22.705    male         0     no  northwest
4      32  28.880    male         0     no  northwest
...   ...     ...     ...       ...    ...        ...
1333   50  30.970    male         3     no  northwest
1334   18  31.920  female         0     no  northeast
1335   18  36.850  female         0     no  southeast
1336   21  25.800  female         0     no  southwest
1337   61  29.070  female         0    yes  northwest

[1338 rows x 6 columns]

In [ ]:
target

charges
0     16884.92400
1      1725.55230
2      4449.46200
3     21984.47061
4      3866.85520
...           ...
1333  10600.54830
1334   2205.98080
1335   1629.83350
1336   2007.94500
1337  29141.36030

[1338 rows x 1 columns]

**One Hot Encoding**
one hot encoding allows the representation  of categorical data  to be more expressive and converted into numerical numbers.One hot encoding ensures that machine learning does not assume that higher numbers are more important.That's why it's preferable.

In [ ]:
from sklearn.preprocessing import LabelEncoder,OneHotEncoder


In [ ]:
encoder=OneHotEncoder(sparse=False)
encoded_labels = pd.DataFrame (encoder.fit_transform(feature[['sex','smoker','region']]))


In [ ]:
encoded_labels.columns = encoder.get_feature_names(['sex','smoker','region'])
dataset= pd.concat([feature, encoded_labels ], axis=1)


In [ ]:
dataset


age     bmi     sex  ...  region_northwest region_southeast region_southwest
0      19  27.900  female  ...               0.0              0.0              1.0
1      18  33.770    male  ...               0.0              1.0              0.0
2      28  33.000    male  ...               0.0              1.0              0.0
3      33  22.705    male  ...               1.0              0.0              0.0
4      32  28.880    male  ...               1.0              0.0              0.0
...   ...     ...     ...  ...               ...              ...              ...
1333   50  30.970    male  ...               1.0              0.0              0.0
1334   18  31.920  female  ...               0.0              0.0              0.0
1335   18  36.850  female  ...               0.0              1.0              0.0
1336   21  25.800  female  ...               0.0              0.0              1.0
1337   61  29.070  female  ...               1.0              0.0              0.0

[1338 rows x 14 columns]

In [ ]:
dataset.columns


Index(['age', 'bmi', 'sex', 'children', 'smoker', 'region', 'sex_female',
       'sex_male', 'smoker_no', 'smoker_yes', 'region_northeast',
       'region_northwest', 'region_southeast', 'region_southwest'],
      dtype='object')

In [ ]:
new_features = dataset[['age', 'bmi', 'children','sex_female', 'sex_male','smoker_no', 'smoker_yes', 'region_northeast','region_northwest', 'region_southeast', 'region_southwest']]


In [ ]:
from sklearn.model_selection import train_test_split

"""Spliting the Dataset into Training Set and Test Set """
X_train,X_test,y_train,y_test=train_test_split(new_features,target,test_size = 0.3,random_state = 0)
# random_state = 0 is select to get the same result


In [ ]:
print(X_train.shape)
print(X_test.shape)


(936, 11)
(402, 11)


In [ ]:
X_train


age     bmi  ...  region_southeast  region_southwest
1163   18  28.215  ...               0.0               0.0
196    39  32.800  ...               0.0               1.0
438    52  46.750  ...               1.0               0.0
183    44  26.410  ...               0.0               0.0
1298   33  27.455  ...               0.0               0.0
...   ...     ...  ...               ...               ...
763    27  26.030  ...               0.0               0.0
835    42  35.970  ...               1.0               0.0
1216   40  25.080  ...               1.0               0.0
559    19  35.530  ...               0.0               0.0
684    33  18.500  ...               0.0               1.0

[936 rows x 11 columns]

In [ ]:
# Fitting Multiple Linear Regression Model to Training Set
from sklearn.linear_model import LinearRegression

regressor = LinearRegression()
regressor.fit(X_train,y_train)

# Predicting the Test Set Results
y_pred = regressor.predict(X_test)


In [ ]:
y_test


charges
578    9724.53000
610    8547.69130
569   45702.02235
1034  12950.07120
198    9644.25250
...           ...
1261   3277.16100
494   17942.10600
97    10226.28420
418   14418.28040
920   13451.12200

[402 rows x 1 columns]

In [ ]:
y_pred


array([[ 1.12531936e+04],
       [ 9.54490709e+03],
       [ 3.78498010e+04],
       [ 1.60692697e+04],
       [ 6.73440872e+03],
       [ 4.00446268e+03],
       [ 1.60576551e+03],
       [ 1.43015300e+04],
       [ 8.89696521e+03],
       [ 7.34111088e+03],
       [ 4.67363726e+03],
       [ 1.01633582e+04],
       [ 8.89221066e+03],
       [ 3.78513354e+03],
       [ 2.74730092e+04],
       [ 1.08538364e+04],
       [ 1.10975451e+04],
       [ 6.15331614e+03],
       [ 8.05810804e+03],
       [ 2.69726437e+04],
       [ 3.33303371e+04],
       [ 1.42507758e+04],
       [ 1.18538537e+04],
       [ 3.19917201e+04],
       [ 4.08938830e+03],
       [ 9.31679198e+03],
       [ 1.14341410e+03],
       [ 9.91266828e+03],
       [ 3.75184042e+03],
       [ 1.02673370e+04],
       [ 8.90007511e+03],
       [ 3.99578709e+04],
       [ 1.57594956e+04],
       [ 1.39309106e+04],
       [ 2.46369056e+04],
       [ 5.22461192e+03],
       [ 1.26648825e+04],
       [ 3.06140223e+04],
       [ 3.3

Evaluation Metrices


In [ ]:
from sklearn.metrics import mean_absolute_error

# MAE

mean_absolute_error(y_test, y_pred)


4011.4496793279886

In [ ]:
from sklearn.metrics import mean_squared_error
mean_squared_error(y_test, y_pred)


33342497.82695459

In [ ]:
from sklearn.metrics import r2_score
r2_score(y_test, y_pred)


0.7909160991789904

Feature Scaling

In [ ]:
dataset


age     bmi     sex  ...  region_northwest region_southeast region_southwest
0      19  27.900  female  ...               0.0              0.0              1.0
1      18  33.770    male  ...               0.0              1.0              0.0
2      28  33.000    male  ...               0.0              1.0              0.0
3      33  22.705    male  ...               1.0              0.0              0.0
4      32  28.880    male  ...               1.0              0.0              0.0
...   ...     ...     ...  ...               ...              ...              ...
1333   50  30.970    male  ...               1.0              0.0              0.0
1334   18  31.920  female  ...               0.0              0.0              0.0
1335   18  36.850  female  ...               0.0              1.0              0.0
1336   21  25.800  female  ...               0.0              0.0              1.0
1337   61  29.070  female  ...               1.0              0.0              0.0

[1338 rows x 14 columns]

In [ ]:
from sklearn.preprocessing import StandardScaler
X_sc = StandardScaler()
y_sc = StandardScaler()
X_train_new = X_sc.fit_transform(X_train[['age', 'bmi', 'children','sex_female', 'sex_male','smoker_no', 'smoker_yes', 'region_northeast','region_northwest', 'region_southeast', 'region_southwest']])
X_test_new = X_sc.transform(X_test[['age', 'bmi', 'children','sex_female', 'sex_male','smoker_no', 'smoker_yes', 'region_northeast','region_northwest', 'region_southeast', 'region_southwest']])
y_train_new = y_sc.fit_transform(y_train[['charges']])
y_test_new = y_sc.transform(y_test[['charges']])


**Support Vector Regression**


In [ ]:
# Fitting SVR to the dataset
from sklearn.svm import SVR 

regressor = SVR(kernel = 'linear')
regressor.fit(X_train_new,y_train_new)



SVR(kernel='linear')

In [ ]:
y_pred = regressor.predict(X_test_new)


In [ ]:
r2_score(y_test_new, y_pred)


0.7619821999423557

**Decision Tree Regression**


In [ ]:
# Fitting the decision tree regression with the dataset
from sklearn.tree import DecisionTreeRegressor

regressor = DecisionTreeRegressor(random_state = 0)
regressor.fit(X_train_new,y_train_new) 


DecisionTreeRegressor(random_state=0)

In [ ]:
y_pred = regressor.predict(X_test_new)
r2_score(y_test_new, y_pred)


0.7154818275740042

Random  Forest Regression


In [ ]:
# Fitting the Random Forest Regression with the dataset
from sklearn.ensemble import RandomForestRegressor

regressor = RandomForestRegressor(n_estimators = 10,random_state = 0) # n estiamator is the number of decision trees
regressor.fit(X_train_new,y_train_new) 


RandomForestRegressor(n_estimators=10, random_state=0)

In [ ]:
y_pred = regressor.predict(X_test_new)
r2_score(y_test_new, y_pred)


0.8402053843182806

It is seen that the highest accuracy is found using Random Forest Regression which is 84%